lib/SILOptimizer/LoopTransforms/LICM.cpp

```diff
@@ -737,14 +819,195 @@ bool LoopTreeOptimization::optimizeLoop(
+  do {
+    changed = false;
+    for (SILBasicBlock *block : loop->blocks()) {
+      bool storeAlive = (storesNotAlive.count(block) == 0);
+      if (storeAlive && stores.count(block) == 0 &&
+          std::any_of(block->pred_begin(), block->pred_end(),
+            [&](SILBasicBlock *b) { return storesNotAlive.count(b) != 0; })) {
+        storesNotAlive.insert(block);
+        changed = true;
+      }
+    }
+  } while (changed);
+
+  auto isUnreachableBlock = [](SILBasicBlock *succ) {
+    return isa<UnreachableInst>(succ->getTerminator());
+  };
+
+  // Check if the store-is-not-alive flag reaches any of the exits.
+  for (SILBasicBlock *eb : exitingBlocks) {
+    // Ignore loop exits to blocks which end in an unreachable.
+    if (!std::any_of(eb->succ_begin(), eb->succ_end(), isUnreachableBlock) &&
+        storesNotAlive.count(eb) != 0) {
+      return false;
+    }
+  }
+  return true;
+}
+
+void LoopTreeOptimization::hoistLoadsAndStores(SILValue addr, SILLoop *loop, InstVector &toDelete) {
+
+  SmallVector<SILBasicBlock *, 4> exitingBlocks;
+  loop->getExitingBlocks(exitingBlocks);
+
+  // This is not a requirement for functional correctness, but we don't want to
+  // _speculatively_ load and store the value (outside of the loop).
+  if (!storesCommonlyDominateLoopExits(addr, loop, exitingBlocks))
+    return;
+
+  // Inserting the stores requires the exit edges to be not critical.
+  for (SILBasicBlock *exitingBlock : exitingBlocks) {
+    for (unsigned idx = 0, e = exitingBlock->getSuccessors().size();
+         idx != e; ++idx) {
+      // exitingBlock->getSuccessors() must not be moved out of this loop,
+      // because the successor list is invalidated by splitCriticalEdge.
+      if (!loop->contains(exitingBlock->getSuccessors()[idx])) {
+        splitCriticalEdge(exitingBlock->getTerminator(), idx, DomTree, LoopInfo);
+      }
+    }
+  }
+
+  SILBasicBlock *preheader = loop->getLoopPreheader();
+  assert(preheader && "Expected a preheader");
+
+  // Initially load the value in the loop pre header.
+  SILBuilder B(preheader->getTerminator());
+  auto *initialLoad = B.createLoad(preheader->getTerminator()->getLoc(), addr,
+                                   LoadOwnershipQualifier::Unqualified);
+
+  SILSSAUpdater ssaUpdater;
+  ssaUpdater.Initialize(initialLoad->getType());
+  ssaUpdater.AddAvailableValue(preheader, initialLoad);
+
+  // Set all stored values as available values in the ssaUpdater.
+  // If there are multiple stores in a block, only the last one counts.
+  Optional<SILLocation> loc;
+  for (SILInstruction *I : LoadsAndStores) {
+    if (auto *SI = isStoreToAddr(I, addr)) {
+      loc = SI->getLoc();
+
+      // If a store just stores the loaded value, bail. The operand (= the load)
+      // will be removed later, so it cannot be used as available value.
+      // This corner case is suprisingly hard to handle, so we just give up.
+      if (isLoadFromAddr(dyn_cast<LoadInst>(SI->getSrc()), addr))
```

Comment:
I guess this `isLoadFromAddr` was a bug that you found after the first implementation, but I really don't get it. Just remove the store if it's not changing the available value! The SSA updater should get it right if it doesn't see this store as a def.